src/DynamicNLPModels.jl

```diff
@@ -576,6 +589,29 @@ function _build_condensed_blocks(
   
     c0 = h0[1,1] / 2
   
+    if S != zeros(size(Q, 1), size(R, 1))
+        block_S = zeros(ns * (N + 1), nu * N)
+
+        for i in 1:N
+            row_range = (1 + ns * (i - 1)):(ns * i)
+            col_range = (1 + nu * (i - 1)):(nu * i)
+
+            block_S[row_range, col_range] = S
+        end        
+
+        BTS      = zeros(nu * N, nu * N)
+        STB      = zeros(nu * N, nu * N)
+        s0T_AT_S = zeros(1, nu * N)
+
+        LinearAlgebra.mul!(BTS, transpose(block_B), block_S)
+        LinearAlgebra.mul!(STB, transpose(block_S), block_B)
+
+        LinearAlgebra.axpy!(1, BTS, BQB)
+        LinearAlgebra.axpy!(1, STB, BQB)
```

Comment:
is `STB` a transpose of `BTS`? If so, no need to do the computation twice, but just use `BTS'` in place of `STB`